File: leetcode/tree/3157_Find_the_Level_of_Tree_with_Minimum_Sum/codes/tree.py
from collections import deque
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def minimumLevel(self, root: Optional[TreeNode]) -> int:
        q = deque([root])
        level = 1
        min_sum = float('inf')
        ans = 1
        while q:
            s = 0
            for _ in range(len(q)):
                node = q.popleft()
                s += node.val
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            if s < min_sum:
                min_sum = s
                ans = level
            level += 1
        return ans

File: leetcode/tree/3157_Find_the_Level_of_Tree_with_Minimum_Sum/codes/tree.c
/**
Definition for a binary tree node.
 */
#include <limits.h>
#include <stdlib.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

struct QueueNode {
    struct TreeNode* node;
    struct QueueNode* next;
};

struct Queue {
    struct QueueNode* front;
    struct QueueNode* rear;
};

void enqueue(struct Queue* q, struct TreeNode* node) {
    struct QueueNode* newNode = (struct QueueNode*)malloc(sizeof(struct QueueNode));
    newNode -> node = node;
    newNode -> next = NULL;
    if (!q -> rear) {
        q -> front = q -> rear = newNode;
    } else {
        q -> rear -> next = newNode;
        q -> rear = newNode;
    }
}

struct TreeNode* dequeue(struct Queue* q) {
    if (!q -> front) return NULL;
    struct QueueNode* temp = q -> front;
    struct TreeNode* node = temp -> node;
    q -> front = q -> front -> next;
    if (!q -> front) q -> rear = NULL;
    free(temp);
    return node;
}

int queueSize(struct Queue* q) {
    int count = 0;
    struct QueueNode* cur = q -> front;
    while (cur) {
        count++;
        cur = cur -> next;
    }
    return count;
}

int minimumLevel(struct TreeNode* root) {
    struct Queue q = {NULL, NULL};
    enqueue(&q, root);
    int level = 1, ans = 1;
    long long minSum = LLONG_MAX;
    while (q.front) {
        int size = queueSize(&q);
        long long sum = 0;
        for (int i = 0; i < size; i++) {
            struct TreeNode* node = dequeue(&q);
            sum += node -> val;
            if (node -> left) enqueue(&q, node -> left);
            if (node -> right) enqueue(&q, node -> right);
        }
        if (sum < minSum) {
            minSum = sum;
            ans = level;
        }
        level++;
    }
    return ans;
}
